src/rules/ResourceICUPlurals.js

```diff
@@ -66,98 +66,200 @@ class ResourceICUPlurals extends ResourceRule {
-                        rule: this,
-                        description: `Extra categories in target string: ${extras.join(", ")}. Expecting only these: ${neededCategories.join(", ")}`,
-                        id: key,
-                        highlight: `Target: ${stringToCheck}<e0></e0>`,
-                        pathName,
-                        source,
-                        locale
-                    };
-                    value.push(new Result(opts));
+    matchCategories(sourceSelect, targetSelect, locale, resource) {
+        let problems = [];
+        const srcLocale = new Locale(resource.getSourceLocale());
+
+        // first check the required plural categories
+        let requiredSourceCategories, requiredTargetCategories;
+        let actualRequiredSourceCategories = [], actualRequiredTargetCategories = [];
+        let actualNonrequiredSourceCategories = [], actualNonrequiredTargetCategories = [];
+
+        if (sourceSelect.node.pluralType === "cardinal") {
+            requiredSourceCategories = categoriesForLang[srcLocale.getLanguage()] || [ "one", "other" ]
+            requiredTargetCategories = categoriesForLang[locale.getLanguage()] || [ "one", "other" ];
+        } else {
+            // for select or selectordinal, only the "other" category is required
+            requiredSourceCategories = [ "other" ];
+            requiredTargetCategories = [ "other" ];
+        }
+
+        for (let category in sourceSelect.node.options) {
+            if (requiredSourceCategories.indexOf(category) > -1) {
+                actualRequiredSourceCategories.push(category);
+            } else {
+                actualNonrequiredSourceCategories.push(category);
+            }
+        }
+
+        for (let category in targetSelect.node.options) {
+            if (requiredTargetCategories.indexOf(category) > -1) {
+                actualRequiredTargetCategories.push(category);
+            } else {
+                actualNonrequiredTargetCategories.push(category);
+            }
+        }
+
+        let missing = requiredTargetCategories.filter(category => {
+            if (!targetSelect.node.options[category]) {
+                // if the required category doesn't exist in the target, check if it is required
+                // in the source. If it is required in the source and does not exist there, then
+                // there is an error in the source and we don't need to register a result for the
+                // target -- the source needs to be fixed first. If it is required in the source
+                // and exists in the source, then produce a result because it should also be there
+                // in the target. If it is not required in the source, then produce a result because
+                // it is required in the target language and it doesn't matter about the source
+                // language.
+                if (requiredSourceCategories.indexOf(category) < 0 || sourceSelect.node.options[category]) {
```

Comment:
2nd check is ambiguous between empty string and undefined

Reply:
In the AST, `sourceSelect.node.options[category]` is either an object or it is undefined. If it is a falsy string, that's the same as undefined, because we can't use it anyways.